fsm/fsm.go

```diff
@@ -31,6 +31,7 @@ type FSM struct {
 	log           zerolog.Logger
 	states        []State
 	transitions   _TransitionMap
+	stater        Stater
```

Comment:
This was what prodded me to think about the `JSONEncoder` interface above: why should the runtime contain some information about serializing the state of the FSM?  There's a hook that's required for metadata, but that should be called at the time that `JSONEncode()` is called, like:

```go
type StateStatusFunc func(w io.Writer) error

func JSONEncoder(fsm *FSM, fn StateStatusFunc, w io.Writer) error {...}
```

?